include/mxnet/ndarray.h

```diff
@@ -421,15 +394,25 @@ class NDArray {
     }
   };
 
+  void SetTBlob() const {
+    tblob_.dptr_ = static_cast<char*>(ptr_->shandle.dptr) + byte_offset_;
+    tblob_.shape_ = shape_;
+    tblob_.type_flag_ = dtype_;
+    tblob_.SetDLTensor(ptr_->shandle.ctx.dev_mask(), ptr_->shandle.ctx.dev_id);
+#if MKL_EXPERIMENTAL == 1
+    tblob_.Mkl_mem_ = Mkl_mem_;
+#endif
+  }
+
 #if MKL_EXPERIMENTAL == 1
   std::shared_ptr<MKLMemHolder> Mkl_mem_;
 #endif
   /*! \brief internal data of NDArray */
   std::shared_ptr<Chunk> ptr_;
   /*! \brief shape of current NDArray */
   TShape shape_;
-  /*! \brief offset in chunk */
-  size_t offset_;
+  /*! \brief byte offset in chunk */
+  size_t byte_offset_ = 0;
```

Comment:
there is no need to switch, the only problem is slice, which is rare. However, there are more conventions that follows byte_offset_ like in DLPack. 

In the future, you should be able derive the byte cost from type(e.g. DLType), removing the need of switch.